dev-docs: add PR review policy (#2349)
canonical/ubuntu-advantage-client: dev-docs/references/pr_review_policy.md

# PR review policy

The team follows this policy for PR reviews:

* Simple PRs: PRs that fix typos or perform one line changes require **one** approval from a team

>   member and can be merged immediately after the review.

Reviewer: I'd specify that the approver himself can merge it

Author: All of those PRs can be merged by anyone in the team after we wait for the right amount of time and have the necessary number of reviews. But I don't know if we should make that explicit on the document.

What do you think @renanrodrigo @orndorffgrant @CalvoM @s-makin ?

Reviewer 2: While we're writing things like this down, I think we might as well be explicit - maybe something like:
```diff
 * Simple PRs: PRs that fix typos or perform one line changes require **one** approval from a team
   member and can be merged immediately by the reviewer. 
 * Low complexity PRs: For short PRs we require **one** approval to merge. However, after approving the
   PR we will wait for **one** day before merging it. This will allow other team members to review
-  the PR if they want.
+  the PR if they want. When one day has passed after the approval and there are no new requested changes,
+  any team member can merge.
 * Complex PRs: Minimum of **two** approvals from team members. Additionally, some PRs may also
   require review from members that are not on the team, like SRU reviewers. However, this is context
-  dependent, as some PRs do not need this type of review.
+  dependent, as some PRs do not need this type of review. After there are two team-member approvals
+  additional review feedback has been taken into account, and there are now remaining requested
+  changes, any team member can merge.
```

Reviewer 3: I think being explicit is always a good idea. The bit about low-complexity PRs gave me a bit of a pause because it looks like we're conflating size with complexity, but the two may or may not be linked (you can have a short/complex or a large/straightforward one). 

It's also a bit unclear from "some PRs do not need this type of review" what PRs that applies to - e.g. does it refer to just the low-complexity ones, or are there special cases of complex ones that don't require it? How do you know which ones need SRU review? Does the team assign reviewers or is it something the proposer has to consider and know who to request review from?

Author: Yes, this is not related to the size at all, but what the PR is actually changing.

And regarding to complex PRs, those are ones that either change/add a Pro service, such as a daemon or a new systemd unit, or ones that are adding features that are tied to system that our team is not an expert at, adding new APT hooks for example. When that happens, we directly ask for review from developers outside the team.

I think I can add some of those examples on that policy

Author: Actually, I have created a new entry for those PRs were we require review from members outside the team. Let me know what you think @orndorffgrant @s-makin

# PR review policy

The team follows this policy for PR reviews:

* Simple PRs: PRs that fix typos or perform one line changes require **one** approval from a team
  member and can be merged immediately by the reviewer.
* Low complexity PRs: For short PRs we require **one** approval to merge. However, after approving the
  PR we will wait for **one** day before merging it. This will allow other team members to review
  the PR if they want. When one day has passed after the approval and there are no new requested
  changes, any team member can merge.
* Complex PRs: Minimum of **two** approvals from team members. After there are two team-member approvals

>   and there are now remaining requested changes, any team member can merge.

Reviewer 3: ```suggestion
  and there are no remaining requested changes, any team member can merge.
```

> * SRU blocker PRs: Some PRs may also require review from members that are not on the team (For example, adding/creating new systemd
>   units), like SRU reviewers. For those PRs, we will directly ask for feedback from member outside

Reviewer 3: ```suggestion
* SRU blocker PRs: Some PRs may also require review from members that are not on the team, like SRU reviewers. This is the case, for example, when adding/creating new systemd units). For those PRs, we will directly ask for feedback from members outside
```

  the team that have a more suitable background to review the change we are proposing. Therefore,

>   not only we require at least **two** approvals from team members, but also approvals for every member
>   outside the team we request the review from. After there are two team-member and outside reviewers approvals, and
>   additional review feedback has been taken into account, and there are now remaining requested changes, any team member can merge.

Reviewer 3: ```suggestion
  we not only require at least **two** approvals from team members, but also approvals from every external reviewer we assigned. After the required approvals have been given, the additional review feedback has been taken into account, and there are no requested changes remaining, any team member can merge.
```
